kernel/src/grant.rs

```diff
@@ -240,75 +449,205 @@ impl<'a> GrantUpcallTable<'a> {
+                // of the ReadOnlyProcessBuffer is bound to the lifetime of self, which
+                // correctly limits dereferencing this saved pointer to only when it is valid.
+                unsafe {
+                    Ok(ReadOnlyProcessBufferRef::new(
+                        saved_ro.ptr,
+                        saved_ro.len,
+                        self.process.processid(),
+                    ))
+                }
+            },
+        )
+    }
+
+    /// Returns a lifetime limited reference to the requested ReadWriteProcessBuffer
+    ///
+    /// The ReadWriteProcessBuffer is only value for as long as this object is valid, i.e.
+    /// the lifetime of the app enter closure.
+    ///
+    /// If the specified allow number is invalid, then a AddressOutOfBounds will be
+    /// return. This returns a process::Error to allow for easy chaining of this
+    /// function with the ReadWriteProcessBuffer::enter function with `and_then`.
+    pub fn get_readwrite_processbuffer(
+        &self,
+        allow_rw_num: usize,
+    ) -> Result<ReadWriteProcessBufferRef, crate::process::Error> {
+        self.allow_rw.get(allow_rw_num).map_or(
+            Err(crate::process::Error::AddressOutOfBounds),
+            |saved_rw| {
+                // # Safety
+                //
+                // This is the saved process buffer data has been validated to be wholly
+                // contained within this process before it was stored. The lifetime
+                // of the ReadWriteProcessBuffer is bound to the lifetime of self, which
+                // correctly limits dereferencing this saved pointer to only when it is valid.
+                unsafe {
+                    Ok(ReadWriteProcessBufferRef::new(
+                        saved_rw.ptr,
+                        saved_rw.len,
+                        self.process.processid(),
+                    ))
+                }
+            },
+        )
+    }
 }
 
 /// A minimal representation of an upcall, used for storing an upcall
 /// in a process' grant table without wasting memory duplicating information
 /// such as process ID.
 #[repr(C)]
-pub(crate) struct SavedUpcall {
-    pub(crate) appdata: usize,
-    pub(crate) fn_ptr: Option<NonNull<()>>,
+struct SavedUpcall {
+    appdata: usize,
+    fn_ptr: Option<NonNull<()>>,
 }
 
-/// Subscribe to an upcall by saving the upcall in the grant region for the
-/// process and returning the existing upcall for the same UpcallId.
-pub(crate) fn subscribe(
-    process: &dyn Process,
-    upcall: Upcall,
-) -> Result<Upcall, (Upcall, ErrorCode)> {
-    let grant_num = match process.lookup_grant_from_driver_num(upcall.upcall_id.driver_num) {
-        Ok(grant_num) => grant_num,
-        Err(e) => return Err((upcall, e.into())),
-    };
+impl Default for SavedUpcall {
+    fn default() -> Self {
+        Self {
+            appdata: 0,
+            fn_ptr: None,
+        }
+    }
+}
```

Comment:
Why not `#[derive(Default)]` on `SavedUpcall`?

Reply:
Done. I do miss the triple symmetric of all three of three `Saved` structures having to implement `Default though :)